src/main/java/software/amazon/smithy/gradle/SmithyPlugin.java

```diff
@@ -1,254 +1,286 @@
-                        throw new GradleException("Detected that this is the main Smithy repo. "
-                                                  + "You need to add an explicit :project dependency on :smithy-cli");
-                    }
-                }
-            }
+    private static void validateExtension(SmithyExtension smithyExtension) {
+        // Set source projection settings
+        // We can only set source projections for projections other than source.
+        if (!smithyExtension.getProjectionSourceTags().isPresent()
+                && smithyExtension.getSourceProjection().get().equals("source")) {
+            throw new GradleException("Projection source tags set with source projection. "
+                    + "Set a value for the `sourceProjection` that matches a projection defined "
+                    + "in one of the smithy-build configs for this project.");
         }
     }
 
-    // Check if there's a dependency on smithy-model somewhere, and assume that version.
-    private String detectCliVersionInDependencies(Configuration configuration) {
-        return configuration.getAllDependencies().stream()
-                .filter(d -> isMatchingDependency(d, "smithy-model"))
-                .map(Dependency::getVersion)
-                .findFirst()
-                .orElse(null);
+
+    /**
+     * Registers a custom Smithy source set as an extension of a given source set.
+     *
+     * <p>Smithy models can be placed in {@code model/}, {@code src/$sourceSetName/smithy},
+     * {@code src/$sourceSetName/smithy}, {@code src/$sourceSetName/resources/META-INF/smithy}, and
+     * {@code src/$sourceSetName/resources/META-INF/smithy}. This code will add these
+     * directories to the specified source set as a "smithy" extension.
+     *
+     * @param sourceSet root sourceSet to add extensions to
+     */
+    private SmithySourceDirectorySet registerSourceSets(SourceSet sourceSet) {
+        // Add the smithy source set as an extension
+        SmithySourceDirectorySet sds = smithyExtension.getSourceSets().create(sourceSet.getName());
+        sourceSet.getExtensions().add(SmithySourceDirectorySet.NAME, sds);
+        SmithySourceDirectorySet.SOURCE_DIRS.forEach(sourceDir ->
+                sds.srcDir(sourceDir.replace("$name", sourceSet.getName())));
+        sds.include("**/*.smithy");
+
+        return sds;
     }
 
-    private static boolean isMatchingDependency(Dependency dependency, String name) {
-        return Objects.equals(dependency.getGroup(), "software.amazon.smithy") && dependency.getName().equals(name);
+    /**
+     * Creates a 'smithyBuild' configuration for the given source set.
+     *
+     * <p>This configuration is used to configure the model discovery classpath and users can
+     * configure dependencies on this configuration that they want included during the
+     * build step but not included as dependencies of their jars.
```

Comment:
This seemed extra wordy, maybe something more direct:
```suggestion
     * <p>This configures the model discovery classpath. Users can set dependencies
     * on this configuration that they want included during the build step, but not
     * included as dependencies of their jars.
```
Maybe it could used more of an explanation of what the model discovery classpath does.